### ISSUE TITLE
Fused AdamW has worse loss than Apex and unfused AdamW for fp16/AMP

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#96755]
### 🐛 Describe the bug

@stas00 recently benchmarked on an 80GB A100 and discovered fused AdamW is worse than both the Apex equivalent and the default implementation. Moreover, fused AdamW is not more performant in this case.
 
| Variation                 |     Train samples per second |   Diff % |   Train loss |
|:--------------------------|----------:|-------:|--------:|
| --optim adamw_torch_fused |    389.41 |      0 |    2.66 |
| --optim adamw_torch       |    389.37 |      0 |    2.55 |
| --optim adamw_apex_fused  |    399.27 |      3 |    2.53 |

Repro:
See @stas00's comment below!
The batch size (--per_device_train_batch_size) can be adjusted to be smaller.

cc @ezyang @gchanan @zou3519 @vincentqb @jbschlosser @albanD @crcrpar

### Versions

Master/latest nightly

**Comment by stas00** [comment]
a few small corrections. 

1. This is only for fp16/amp - bf16/amp and fp32 perform correctly
2. I think I may have missed `--fp16` in the repro, here is the correct one:

```
git clone https://github.com/huggingface/transformers
cd transformers
pip install -e .[torch]
git checkout adamw_fused_torch
PYTHONPATH=src CUDA_VISIBLE_DEVICES=0 python scripts/benchmark/trainer-benchmark.py --base-cmd ' \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir \
--do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 \
--max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: "  --warmup_steps 50 \
--max_train_samples 20000 --dataloader_num_workers 2 --fp16 \
' --target-metric-key train_samples_per_second --repeat-times 1 --variations '--optim adamw_torch_fused|--optim adamw_torch|--optim adamw_apex_fused' --report-metric-keys train_loss --base-variation '--optim adamw_torch'

```

**Comment by crcrpar** [comment]
Thank you for reporting

@stas00 would you know if this is consistently reproducible with smaller "steps"?

@janeyx99 just to confirm, `torch.optim.AdamW.step` by default calls `_multi_tensor_adamw`, right?

**Comment by stas00** [comment]
do you mean a smaller lr? by default it used `--learning_rate 5e-5`

please suggest which one would you like to try, you can just add `--learning_rate 5e-5` to the base command argument

**Comment by janeyx99** [comment]
> @janeyx99 just to confirm, `torch.optim.AdamW.step` by default calls `_multi_tensor_adamw`, right?

Oh yes right, updated issue body to say the default, which should be foreach and not the for-loop implementation.

**Comment by crcrpar** [comment]
excuse me for the delay, was trying to narrow down as possible.

By steps, I meant the number of parameter updates following the command line argument of `--max_steps`. sorry for the confusion.
for now I'm able to see some difference between fused and foreach with smaller number of training samples (approximately ~10 optimizer.step calls after the smaller init_scale for GradScaler) but am yet to figure out the cause. Currently I somehow see the different number of `optimizer.step` being skipped

**Comment by stas00** [comment]
Please do let me know if I can support this process in any way. Are you able to run the benchmarking code I shared? So that you're in control and can change the knobs as you see fit.

By skipping do you mean when it hits overflow and the loss scaler is scaled down? 

If so does it use some different logic than the non-fused one? I'd imagine they should be matched computation-wise.

**Comment by crcrpar** [comment]
yet to fix it but it seems like the interaction between the fused optimizer and the grad scaler is the major cause.

- the diverged loss: seems to be restored if I disable the interaction, i.e. change [this line](https://github.com/pytorch/pytorch/blob/f865e23abccf41e37f94917c9f518ebac9dc9c90/torch/optim/adamw.py#L55) to False
- the mediocre performance: seems to be related to that many iterations end up with invalid gradients. To be compatible with cuda graph even when GradScaler is used, the fused interacts with it and calls `_fused_adamw` function regardless of the value of `found_inf` while the others not. This causes [redundant parameter grouping](https://github.com/pytorch/pytorch/blob/f865e23abccf41e37f94917c9f518ebac9dc9c90/torch/optim/adamw.py#L594) in `_fused_adamw` and for-loop in `optimizer.step`.

**Comment by crcrpar** [comment]
I ran benchmark of `(foreach, vanilla, fused) x (fp32, bf16, fp16)` for translation and language modeling example because in the former example I saw `GradScaler`'s scale became 0 despite the value of `fused` (scaling value of gradients of 0 basically indicates the training hasn't been working nicely).

I will keep investigating the fused optimizer as the discrepancy is still concerning / unexpected and I cannot explain why.

### Translation Example

Compared to BFloat16, loss values of fp16 are larger than those of bfloat16 and float32, so I'm wondering if this script converges with fp16.

*** Results:

| Variation                             |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused             |                              314.24 |           9 |            2.21 |
| --optim adamw_torch_fused --fp16 True |                              359.36 |          25 |            2.66 |
| --optim adamw_torch_fused --bf16 True |                              363.29 |          26 |            2.22 |
| --optim adamw_torch                   |                              288.49 |           0 |            2.21 |
| --optim adamw_torch --fp16 True       |                              360.51 |          25 |            2.55 |
| --optim adamw_torch --bf16 True       |                              333.11 |          15 |            2.22 |

```

*** Setup:


Datetime    : 2023-03-17 16:51:01

Software:
transformers: 4.28.0.dev0
torch       : 2.1.0a0+git13538c8
cuda        : 12.1
python      : 3.10.9

Hardware:
1 GPUs      : NVIDIA A100 80GB PCIe, 79.18GB


*** The benchmark command line was:

CUDA_VISIBLE_DEVICES=0 python scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir --do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 --max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir --source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" --source_prefix "translate English to Romanian: "  --warmup_steps 50 --max_train_samples 20000 --dataloader_num_workers 32 ' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adamw_torch_fused|--optim adamw_torch' '|--fp16 True|--bf16 True' \
--report-metric-keys train_loss --base-variation '--optim adamw_torch'

```

### Language Modeling Example

In this example, fp16 AMP seems fine according to the rows of `adamw_torch` and `adamw_torch_fused` looks okay to me.
Please correct me if  `--max_train_samples` of 


*** Results:

| Variation                             |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused             |                               28.39 |           8 |            2.97 |
| --optim adamw_torch_fused --fp16 True |                               29.32 |          12 |            2.97 |
| --optim adamw_torch_fused --bf16 True |                               30.86 |          17 |            2.97 |
| --optim adamw_torch                   |                               26.28 |           0 |            2.97 |
| --optim adamw_torch --fp16 True       |                               27.32 |           4 |            2.97 |
| --optim adamw_torch --bf16 True       |                               28.30 |           8 |            2.97 |

```

*** Setup:


Datetime    : 2023-03-17 17:40:10

Software:
transformers: 4.28.0.dev0
torch       : 2.1.0a0+git13538c8
cuda        : 12.1
python      : 3.10.9

Hardware:
1 GPUs      : NVIDIA A100 80GB PCIe, 79.18GB


*** The benchmark command line was:

CUDA_VISIBLE_DEVICES=0 python scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' examples/pytorch/language-modeling/run_clm.py --model_name_or_path gpt2-medium --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --logging_strategy no --save_strategy no --do_train --max_train_samples 2400 --per_device_train_batch_size 4 --num_train_epochs 1 --warmup_steps 8 --block_size 512 --report_to none --output_dir lm_benchmark --overwrite_output_dir ' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adamw_torch_fused|--optim adamw_torch' '|--fp16 True|--bf16 True' \
--report-metric-keys train_loss --base-variation '--optim adamw_torch'

```

**Comment by stas00** [comment]
I see what you mean, you're saying that in general that `lr` doesn't go well for fp16 since torch adam (non fused) fares worse under fp16.

But as you can see from the same training, using a wider range of optimizers:
https://github.com/huggingface/transformers/issues/22101#issue-1619885644
they more or less agree in their loss and so it's only the fused version of torch that has a lower loss. so even if the setup is not optimal I'd expect the same algorithm lead to the same loss regardless if it's fused or not. Would you agree?

I'm glad you figured out my little variations syntax and so the benchmarking script did the work for you. Hope it was easy to use. I think you're the second person to use it.

**Comment by janeyx99** [comment]
@stas00 @crcrpar recently pushed a fix for fp16 that should be in tonight's nightly. It should fix the loss and be a general perf gain overall!

**Comment by stas00** [comment]
Amazing, thank you guys for fixing!

I will retest then with the next nightly!

Is it going to land in pt=2.0.1?

**Comment by janeyx99** [comment]
This is the hope, though discussion on 2.0.1 is still ongoing so I can't give a certain answer yet.

**Comment by stas00** [comment]
I thought it was a certainty, no? At least the original fp16 fix, that you said couldn't enter 2.0.0, right?

we already released transformers that currently disables this optimizer for fp16/amp but it was programmed to enable it for torch>2.0.0 so if possible let's put it in 2.0.1.

**Comment by janeyx99** [comment]
Let me clarify: this change should make it into whatever the next release should be. To my knowledge, the existence of 2.0.1 is likely but not 100% yet.

**Comment by stas00** [comment]
Perfect. We are good then. Thank you for clarifying, Jane.

**Comment by stas00** [comment]
I confirm that the problem has been solved:

| Variation                 |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused |                              387.10 |           3 |            2.66 |
| --optim adamw_torch       |                              377.61 |           0 |            2.66 |
| --optim adamw_apex_fused  |                              389.49 |           3 |            2.66 |

Thank you, @crcrpar

**Comment by janeyx99** [comment]
Woooo!! @crcrpar thank you so much for the speedy fix! :D

**Comment by crcrpar** [comment]
Appreciate the quick confirmation :)

**Comment by janeyx99** [comment]
Reopening as the fix has been reverted for causing hangs on internal gloo models, see https://github.com/pytorch/pytorch/pull/98613.

@stas00 Wanted to notify you that the fix has been reverted and we are renewedly working on fixing it + will let you know once a new fix is out.

**Comment by janeyx99** [comment]
<img width="1252" alt="image" src="https://user-images.githubusercontent.com/31798555/231563653-bab00c13-dc21-4464-92d8-170bb3f555df.png">

After looking more closely at the numbers, @albanD noticed that the fix simply made everything consistent with the (lower) performance of the torch fused adam vs. promoting everything to match the apex perf.

Consistency is good, but this does raise some observations/questions:
1. **Why did apex fused get slower?** I suspect the initial results for apex fused did not use fp16 while the second result does, which would make sense for why the apex fused path lost performance. Is this correct? I would guess that the torch32 fused implementations would still beat the fp16 enabled ones.
2. **How come the old version of unfused torch adamw was faster?** @stas00 Does the hf t5 model do clamping of weights or grads at all? The difference that was rectified was that the non-fused paths now check that the unscaled gradients are valid before passing them to the optimizer. Without this check, I would expect for invalid gradients (nans and infinites) to propagate to parameter weights and eventually the loss, which was NOT happening. Thus, my hypothesis is that these may be dropped or clamped in some way. which would be better than the "safer path".
3. **How come the fused torch adamw with fp16 acts as if it checks the grads after unscaling when the code seems to state otherwise?** This I don't know yet, but @crcrpar may have a clue.

With these observations, I think we should consider two things:
1. We should strive for consistency, meaning we should figure out the discrepancies between all the grad scaler codepaths and make their accuracies/functionalities equal. It looks like there is still some unsolved discrepancy here with the grad scaler regarding fused + unfused, and we should eradicate that.
2. We should decide on which path to go on. The numbers say that it is better to align with the nonfused + don't check for nonsense on unscaled gradients. But perhaps it would be safer to add the check anyway and disallow the nonsensical gradients from propagating to the model with a warning at the cost of perf. This could be a way we tell users that the model params may simply not be suitable for amp with fp16. This decision likely depends on resolving question 3 from above.

**Comment by stas00** [comment]
Wow, thank you for catching this @albanD and @janeyx99 - this is bad indeed!

So I rerun the tests, w/ pt-2.0:


*** Results:

| Variation                 |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused |                              357.31 |          -6 |            0.00 |
| --optim adamw_torch       |                              382.13 |           0 |            2.55 |
| --optim adamw_apex_fused  |                              393.52 |           3 |            2.53 |

and w/ pt-2.1-nightly:

| Variation                 |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused |                              384.63 |          25 |            2.66 |
| --optim adamw_torch       |                              308.65 |           0 |            2.66 |
| --optim adamw_apex_fused  |                              387.86 |          26 |            2.66 |

How is it even possible that apex produces 2 different results depending on pytorch version.

I should try and rebuild both apex versions to ensure I'm testing the same code.

**Comment by janeyx99** [comment]
> Before I address your commentary - how is it even possible that apex and unfused torch both produce 2 different loss results depending on pytorch version. Some sort of regression in pytorch?

The fix in question modified the grad scaler for the unfused path, which is used whenever amp with fp16 is specified. The fact that the unfused path was changed is thus directly consequential, and it makes sense that in torch the apex path would change too. To give more detail, there are two paths the grad scaler takes when stepping, depending on whether the optimizer itself would handle scaling. The apex + non-fused fall in one (where they do not handle scaling) whereas the fused fall in the other.

How come the loss for adamw_torch_fused is now 0 in 2.0? Isn't that better now/different from the previous 2.66? Is this because some sort of clamping occurred? Instead of comparing against 2.0, could you compare against the nightly on 3/5 after this change? https://github.com/pytorch/pytorch/pull/95847.

**Comment by stas00** [comment]
Are you saying that apex's implementation of FusedAdamW isn't completely standalone and instead it fuses the pytorch implementation?

> How come the loss for adamw_torch_fused is now 0 in 2.0? 

It's totally weird, I was hoping you can tell me why it's 0. Perhaps the loss scaler never fits and it basically always skips all iterations.

Here is the direct run with uncaught output:

w/ pt-2.0.0+cu117:

```
PYTHONPATH=src CUDA_VISIBLE_DEVICES=0 python examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir --do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 --max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir --source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" --source_prefix "translate English to Romanian: "  --warmup_steps 50 --max_train_samples 20000 --dataloader_num_workers 2 --fp16 --optim adamw_torch_fused --logging_steps 5 --logging_strategy steps --learning_rate 5e-6
[...]
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.01}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.02}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.02}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.03}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.04}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.05}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.06}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.06}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.07}
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.08}
``` 

I think the key is that LR is 0.

same with pt-2.1.0.dev20230327+cu117:

```
{'loss': 0.0, 'learning_rate': 0.0, 'epoch': 0.01}                                                                                        
{'loss': 2.38, 'learning_rate': 2.0000000000000002e-07, 'epoch': 0.02}                                                                    
{'loss': 0.8795, 'learning_rate': 3.0000000000000004e-07, 'epoch': 0.02}                                                                  
{'loss': 2.0755, 'learning_rate': 6.000000000000001e-07, 'epoch': 0.03}
{'loss': 2.622, 'learning_rate': 9.000000000000001e-07, 'epoch': 0.04}
{'loss': 3.1298, 'learning_rate': 1.4000000000000001e-06, 'epoch': 0.05}
{'loss': 2.2564, 'learning_rate': 1.7000000000000002e-06, 'epoch': 0.06}
{'loss': 0.8893, 'learning_rate': 1.8000000000000001e-06, 'epoch': 0.06}
```

**Comment by stas00** [comment]
> Instead of comparing against 2.0, could you compare against the nightly on 3/5 after this change? https://github.com/pytorch/pytorch/pull/95847.

w/ pt=2.1.0.dev20230306+cu117

| Variation                 |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused |                              390.51 |           0 |            2.66 |
| --optim adamw_torch       |                              388.95 |           0 |            2.55 |
| --optim adamw_apex_fused  |                              397.17 |           2 |            2.53 |

**Comment by crcrpar** [comment]
> 3\. **How come the fused torch adamw with fp16 acts as if it checks the grads after unscaling when the code seems to state otherwise?** This I don't know yet, but @crcrpar may have a clue.

https://github.com/pytorch/pytorch/blob/bd07f8d2e03d2d53865584d04945c9f36d4be019/torch/cuda/amp/grad_scaler.py#L353 is the inf check of scaled gradients when we call only `GradScaler.step(optimizer)`, but is the inf check of unscaled gradients if we explicitly call `GradScaler.unscale_(optimizer)`.

**Comment by janeyx99** [comment]
Oh! I see! So the nonfused optimizers had guarded the check on the opt_state being READY but the fused path had not, which is only relevant for the direct unscale_ API. This means that @crcrpar's https://github.com/pytorch/pytorch/pull/98620 _should_ fix this discrepancy. @crcrpar would it be possible to get some benchmark results on that PR to confirm? @stas00 could you confirm that there is manual unscaling involved (e.g., there is a direct `scaler.unscale_(...)` call similar to in https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-clipping)?

Regarding the 0 loss, I think that is due to some post processing. The loss becoming NaNs was another [issue](https://github.com/pytorch/pytorch/issues/95781) that was already fixed.

**Comment by crcrpar** [comment]
> The loss becoming NaNs was another [issue](https://github.com/pytorch/pytorch/issues/95781) that was already fixed and will go into 2.0.1.

How is that issue related? I didn't set `fused` to `True` in the log I attached in my last comment.

**Comment by stas00** [comment]
>  @stas00 could you confirm that there is manual unscaling involved (e.g., there is a direct scaler.unscale_(...) call similar to in https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-clipping)?

Yes, it's there:

https://github.com/huggingface/transformers/blob/e13d6ef7dce05316440febbe62c66fbd10b84c11/src/transformers/trainer.py#L1961-L1962

**Comment by janeyx99** [comment]
@crcrpar Maybe the loss is processed by clipping or something like https://github.com/huggingface/transformers/blob/e13d6ef7dce05316440febbe62c66fbd10b84c11/src/transformers/trainer.py#L1937?  I haven't looked too closely, but I suspect there is additional logic or clipping that curbs the loss away from becoming invalid here.

**Comment by janeyx99** [comment]
@stas00 based on @crcrpar's benchmarking results in https://github.com/pytorch/pytorch/pull/98620#issuecomment-1512227931, it looks like the "better" loss would be a nan under the hood without the logging_nan_inf_filter argument.

If it were disabled, both fused and unfused torch (and I suspect apex) would have nan losses instead of "lower" losses. This brings me back to my second question from an earlier comment: which path do we want our implementations to align on? 

If we align on the "check for inf/nans after unscaling the grads," we will settle for safer but seemingly less performant loss. Safety trading off performance makes sense--being safer means skipping failed iterations to find a better scale, essentially "wasting" that iteration compute. I can conceive a situation here where there doesn't exist a scale that works for the model params, and the model may just converge on a higher loss. For this issue, the higher loss of 2.66 is not much worse, but maybe there is a situation where the model may just stop running very early on due to there not being a viable scale.

If we align on the "check for inf/nans BEFORE unscaling the grads," we will be sticking with the original default, which checks the scaled grads and assumes the unscaled grads should be okay. In the cases where the unscaled grads are invalid, this means that we allow users to handle the nans themselves, leading to better looking losses on this particular issue though...are they actually better?

Currently, I am leaning on the sticking to the default (so the second option, and what landing #98620 would be). In the rare occasion where unscaled grads are bad when scaled grads are okay (so when the scale goes extremely small), the losses would be polluted and give users signal more immediately. Then, the users would have the freedom to filter nans as they please or choose to not use amp for that specific model. This would also be "less" BC breaking, since, to my understanding, this is what NVIDIA's already set as standard in the [docs](https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html#scalefactor).

**Comment by stas00** [comment]
In general it feels like this should be taken into its own Issue and discussed with a larger group of people since such choice shouldn't be taken lightly.

> If we align on the "check for inf/nans after unscaling the grads," ...

According to the [doc](https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html#scalefactor) you quoted the scale factor should be 8 to 32K, so I'm not sure how unscaling grads could lead to a nan/inf since unscaling is `1/S`? The nan/inf result after `1/S` could only happen if `0 < S < 1`, but it's not so. `S>8` in the worst case can only lead to 0 grads.

So is the first case even an option to consider?

**Comment by janeyx99** [comment]
Okay, it sounds like we should continue with our default for now (option 2), and create another issue if needed for the bigger decision.

For how the nans/infs could happen--the scale could become very tiny and so g/S becomes infinity. This is what happens in https://gist.github.com/crcrpar/822a55b5709e0c9371b2c21362698107#file-t5_adamw_torch_550_steps-log-L1383 where the scale is tiny, the grad is inf, and the params become nans. The key here could be to prevent scale from ever being less than 1.

**Comment by stas00** [comment]
oh, I see in the log - you're going into `0 < S < 1` range - then yes, it could be a problem, after unscaling.

The NVIDIA doc suggests `S>1`, no?

again, in Deepspeed's implementation of a scaler if `S=1` leads to an overflow the optimizer quits.

Is there a support for going into the `0 < S < 1` range?

--------------------

So if any S is supported perhaps the right solution is:

1. if `S>1` check the scaled grads
2. if `S<1` check the unscaled grads?

**Comment by janeyx99** [comment]
For the sake of clarity and modularity, I believe #98620 addresses the original concern brought up by this issue by making the fused + nonfused paths consistent with prior established behavior of checking scaled grads and not unscaled grads. It should then align in terms of the hf benchmarks to the lower loss numbers when nan filtering is on. Thus, once the PR lands, we can settle and close this issue.

For the bigger question of standardizing what we should do when invalid grads appear, I have opened this new issue https://github.com/pytorch/pytorch/issues/99640

**Comment by janeyx99** [comment]
Closing as the relevant PR has landed. Thanks @crcrpar for the PR!

**Comment by stas00** [comment]
Hi @janeyx99, as discussed in https://github.com/pytorch/pytorch/issues/99640 `t5-base` was a bad model to test the optimizer with under fp16/amp since the model is bf16-pretrained, so it overflows and doesn't help with the benchmark. The optimizer should have failed but it didn't.

So from now on let's switch to a different model that was fp16-pretrained and should be good for testing, so please use this benchmark instead:

```
PYTHONPATH=src CUDA_VISIBLE_DEVICES=0 python scripts/benchmark/trainer-benchmark.py --base-cmd \
' \
examples/pytorch/language-modeling/run_clm.py --model_name_or_path gpt2 --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --logging_strategy no --save_strategy no  --do_train --max_train_samples 10000  --per_device_train_batch_size 16  --num_train_epochs 1 --warmup_steps 8 --block_size 512 --fp16 --report_to none --output_dir output_dir --fp16 \
' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adamw_torch_fused|--optim adamw_torch|--optim adamw_apex_fused' \
--report-metric-keys train_loss --base-variation '--optim adamw_torch'

```

I have run it on the nightly of today:


| Variation                 |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused |                               89.24 |           3 |            3.35 |
| --optim adamw_torch       |                               86.97 |           0 |            3.35 |
| --optim adamw_apex_fused  |                               89.76 |           3 |            3.35 |

Looking good!

**Comment by janeyx99** [comment]
Awesome!! :D thanks @stas00 for verifying :)
